File: 2017/day1/day1.clj
'{:dependencies [[org.clojure/clojure "1.10.1"]]}
(require '[clojure.string :as str])

(def input
  (->>
   (slurp "2017/day1/input.txt")
   (str/trim)
   (#(str/split % #""))
   (map #(Integer/parseInt %))))

; Example input
;; (def input (-> 91212129
;;                (str)
;;                (str/split #"")
;;                ((fn [s] (map #(Integer/parseInt %) s)))))

(defn match-at
  "Check if value matches list at index"
  [list index value]
  (let [length (count list)
        wrapped-index (mod (+ index length) length)]
    (= (nth list wrapped-index) value)))

(defn part1 []
  (prn "day1 part1:"
       (->>
        (map-indexed
         (fn [idx val]
           (when (match-at input (inc idx) val) val)) input)
        (remove nil?)
        (reduce +))))

(defn part2 []
  (prn "day1 part2:"
       (->>
        (map-indexed
         (fn [idx val]
           (when (match-at input (+ idx (/ (count input) 2)) val) val)) input)
        (remove nil?)
        (reduce +))))

(part1)
(part2)

File: 2017/day1/day1.js
const fs = require('fs');
const path = require('path');
const util = require('util');

util.inspect.defaultOptions.maxArrayLength = null;
// util.inspect.defaultOptions.showHidden = true;
// util.inspect.defaultOptions.depth = null;
// util.inspect.defaultOptions.compact = true;

const log = (...args) => console.log(...args);
const sum = (acc, cur) => acc + cur;

function matchAt(list, index, value) {
  index = (index + list.length) % list.length;
  return list[index] === value;
}

function part1(input) {
  let nums = input
    .trim()
    .split('')
    .map(Number);

  return nums
    .map((val, index) => matchAt(nums, index + 1, val) ? val : null)
    .filter(val => !!val)
    .reduce(sum);
}

function part2(input) {
  let nums = input
    .trim()
    .split('')
    .map(Number);

  let halfWayAround = nums.length / 2;

  return nums
    .map((val, index) => matchAt(nums, index + halfWayAround, val) ? val : null)
    .filter(val => !!val)
    .reduce(sum, 0);
}

let input = fs.readFileSync(path.resolve(__dirname, './input.txt'), 'utf8');

log('day1 part1:', part1(input));
log('day1 part2:', part2(input));
